tests/hikari/hikari_test_helpers.py

```diff
@@ -188,3 +188,27 @@ def __call__(self, *args, **kwargs):
         args = (copy.copy(arg) for arg in args)
         kwargs = {copy.copy(key): copy.copy(value) for key, value in kwargs.items()}
         return super().__call__(*args, **kwargs)
+
+
+class AwaitableMock:
```

Comment:
Couldn't you just use AsyncMock and do
```
foo = AsyncMock()
thing_thats_going_to_be_awaited = foo()

# await thing_thats_going_to_be_awaited happens somewhere

bar.assert_awaited_once()
```

and

```
foo = AsyncMock()
thing_thats_not_going_to_be_awaited = foo()

# await thing_thats_not_going_to_be_awaited  shouldn't happen

bar.assert_not_awaited()
```

Reply:
This doesn't work because `mock.AsyncMock` creates a new object every call.

```
E               AssertionError: Calls not found.
E               Expected: [call(<coroutine object AsyncMockMixin._execute_mock_call at 0x7fcca72f1a40>),
E                call(<coroutine object AsyncMockMixin._execute_mock_call at 0x7fcca7273c40>),
E                call(<coroutine object AsyncMockMixin._execute_mock_call at 0x7fcca7273bc0>),
E                call(<coroutine object AsyncMockMixin._execute_mock_call at 0x7fcca7273ec0>),
E                call(<coroutine object AsyncMockMixin._execute_mock_call at 0x7fcca7273e40>)]
E               Actual: [call(<coroutine object AsyncMockMixin._execute_mock_call at 0x7fcca72be1c0>),
E                call(<coroutine object AsyncMockMixin._execute_mock_call at 0x7fcca72be2c0>),
E                call(<coroutine object AsyncMockMixin._execute_mock_call at 0x7fcca72be3c0>),
E                call(<coroutine object AsyncMockMixin._execute_mock_call at 0x7fcca72be340>),
E                call(<coroutine object AsyncMockMixin._execute_mock_call at 0x7fcca72be4c0>)]

```

Reply:
Welp works on my machine unless you're trying to re-use the same AsyncMock instance for different multiple awaitable fields for some reason

Reply:
Ah I you're not actually using this for new test cases, why does this need to be moved to the helpers when it's only for one very specific test case and wouldn't apply anywhere else?

Reply:
```py
    @pytest.mark.asyncio
    async def test_close_when_force(self, bot, event_manager, event_factory, rest, voice, cache):
        def null_call(arg):
            return arg

        stack = contextlib.ExitStack()
        stack.enter_context(mock.patch.object(asyncio, "as_completed", side_effect=null_call))
        ensure_future = stack.enter_context(mock.patch.object(asyncio, "ensure_future", side_effect=null_call))
        get_running_loop = stack.enter_context(mock.patch.object(asyncio, "get_running_loop"))
        all_of = stack.enter_context(mock.patch.object(aio, "all_of", new=mock.AsyncMock()))

        event_manager.dispatch = mock.AsyncMock()
        rest_close_coro = mock.AsyncMock()
        rest.close = mock.Mock(return_value=rest_close_coro())
        voice_close_coro = mock.AsyncMock()
        voice.close = mock.Mock(return_value=voice_close_coro())
        bot._closing_event = closing_event = mock.Mock(is_set=mock.Mock(return_value=False))
        bot._is_alive = True
        error = RuntimeError()
        shard0_coro = mock.AsyncMock()
        shard0 = mock.Mock(id=0, close=mock.Mock(return_value=shard0_coro()))
        shard1_coro = mock.AsyncMock(side_effect=error)
        shard1 = mock.Mock(id=1, close=mock.Mock(return_value=shard1_coro()))
        shard2_coro = mock.AsyncMock()
        shard2 = mock.Mock(id=2, close=mock.Mock(return_value=shard2_coro()))
        bot._shards = {0: shard0, 1: shard1, 2: shard2}

        with stack:
            await bot.close(force=True)

        # Closing event and args
        closing_event.set.assert_called_once_with()
        assert bot._closing_event is None
        assert bot._is_alive is False

        # Closing components
        ensure_future.assert_has_calls(
            [
                mock.call(rest.close.return_value),
                mock.call(voice.close.return_value),
                mock.call(shard0.close.return_value),
                mock.call(shard1.close.return_value),
                mock.call(shard2.close.return_value),
            ],
            any_order=False,
        )

        rest.close.assert_called_once()
        rest_close_coro.assert_awaited_once()
        voice.close.assert_called_once()
        voice_close_coro.assert_awaited_once()
        shard0.close.assert_called_once()
        shard0_coro.assert_awaited_once()
        shard1.close.assert_called_once()
        shard1_coro.assert_awaited_once()
        shard2.close.assert_called_once()
        shard2_coro.assert_awaited_once()

        # Error handling
        get_running_loop.assert_called_once_with()
        get_running_loop.return_value.call_exception_handler.assert_called_once_with(
            {
                "message": "shard 1 raised an exception during shutdown",
                "future": shard1.close.return_value,
                "exception": error,
            }
        )

        # Joining shards
        all_of.assert_awaited_once_with(*(shard0.join(), shard1.join(), shard2.join()), timeout=3)

        # Clear out maps
        assert bot._shards == {}
        cache.clear.assert_called_once_with()

        # Dispatching events in the right order
        event_manager.dispatch.assert_has_calls(
            [
                mock.call(event_factory.deserialize_stopping_event.return_value),
                mock.call(event_factory.deserialize_stopped_event.return_value),
            ],
            any_order=False,
        )

```
also FWIW this works fine for me without the need for a special helper class

Reply:
> Ah I you're not actually using this for new test cases, why does this need to be moved to the helpers when it's only for one very specific test case and wouldn't apply anywhere else?

First iteration of tests actually used it, which is why I moved it. Since I'm not using it anymore, I moved it back. Tests should be fine now